src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/OpenSslCachedSystemStoreProvider.cs

```diff
@@ -126,61 +147,76 @@ public void Remove(ICertificatePal cert)
             SafeX509StackHandle intermedStore = Interop.Crypto.NewX509Stack();
             Interop.Crypto.CheckValidOpenSslHandle(intermedStore);
 
-            DateTime newFileTime = default;
-            DateTime newDirTime = default;
-
             var uniqueRootCerts = new HashSet<X509Certificate2>();
             var uniqueIntermediateCerts = new HashSet<X509Certificate2>();
             bool firstLoad = (s_nativeCollections == null);
 
-            if (rootStoreFile != null && rootStoreFile.Exists)
+            if (firstLoad)
+            {
+                s_rootStoreDirectories = GetRootStoreDirectories(out s_defaultRootDir);
+                s_directoryLastWrite = new DateTime[s_rootStoreDirectories.Length];
+                s_rootStoreFile = GetRootStoreFile();
+            }
+            else
+            {
+                Debug.Assert(s_rootStoreDirectories is not null);
+                Debug.Assert(s_directoryLastWrite is not null);
+            }
+
+            if (s_rootStoreFile != null)
             {
-                newFileTime = ContentWriteTime(rootStoreFile);
-                ProcessFile(rootStoreFile);
+                ProcessFile(s_rootStoreFile, out s_fileLastWrite);
             }
 
             bool hasStoreData = false;
 
-            if (rootStorePath != null && rootStorePath.Exists)
+            for (int i = 0; i < s_rootStoreDirectories.Length; i++)
             {
-                newDirTime = ContentWriteTime(rootStorePath);
-                hasStoreData = ProcessDir(rootStorePath);
+                hasStoreData = ProcessDir(s_rootStoreDirectories[i], out s_directoryLastWrite[i]);
```

Comment:
This updates the array even if it returns `false`. Should this do something like

```suggestion
                hasStoreData = ProcessDir(s_rootStoreDirectories[i], out DateTime localDirectoryLastWrite);
                
                if (hasStoreData)
                {
                    s_directoryLastWrite[i] = localDirectoryLastWrite;
                }
```

Or does `ProcessDir` need to return a value to make a distinction between "I could not stat the directory" vs. "There was nothing in the directory"?